--- build/ext/zlib/zlib-1.2.8.org/CMakeLists.txt	2013-04-29 07:57:10.000000000 +0900
+++ build/ext/zlib/zlib-1.2.8/CMakeLists.txt	2013-09-22 11:33:03.000000000 +0900
@@ -83,7 +83,7 @@
 		${ZLIB_PC} @ONLY)
 configure_file(	${CMAKE_CURRENT_SOURCE_DIR}/zconf.h.cmakein
 		${CMAKE_CURRENT_BINARY_DIR}/zconf.h @ONLY)
-include_directories(${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_SOURCE_DIR})
+include_directories(${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_SOURCE_DIR} ${CMAKE_CURRENT_SOURCE_DIR})
 
 
 #============================================================================
