*** gc-7.2/dyn_load.c	2013-07-26 14:31:51.272450500 +0200
--- gc-7.2/dyn_load.c.qnx	2013-07-26 14:32:20.443783600 +0200
***************
*** 63,69 ****
      !(defined(FREEBSD) && defined(__ELF__)) && \
      !(defined(OPENBSD) && (defined(__ELF__) || defined(M68K))) && \
      !(defined(NETBSD) && defined(__ELF__)) && !defined(HURD) && \
!     !defined(DARWIN) && !defined(CYGWIN32)
   --> We only know how to find data segments of dynamic libraries for the
   --> above.  Additional SVR4 variants might not be too
   --> hard to add.
--- 63,69 ----
      !(defined(FREEBSD) && defined(__ELF__)) && \
      !(defined(OPENBSD) && (defined(__ELF__) || defined(M68K))) && \
      !(defined(NETBSD) && defined(__ELF__)) && !defined(HURD) && \
!     !defined(DARWIN) && !defined(CYGWIN32) && !defined(QNX)
   --> We only know how to find data segments of dynamic libraries for the
   --> above.  Additional SVR4 variants might not be too
   --> hard to add.
