diff --color -ruN ./sagittarius-0.9.14/ext/ffi/sagittarius-ffi.c ./sagittarius-0.9.14-fix/ext/ffi/sagittarius-ffi.c
--- ./sagittarius-0.9.14/ext/ffi/sagittarius-ffi.c	2025-12-25 01:12:19
+++ ./sagittarius-0.9.14-fix/ext/ffi/sagittarius-ffi.c	2026-01-06 09:30:48
@@ -1584,7 +1584,7 @@
     if (!SG_NUMBERP(ret)) goto ret0;
     *((int64_t *) result) = Sg_GetIntegerS64Clamp(ret, SG_CLAMP_NONE, NULL);
     break;
-#if SIZEOF_LONG != 4
+#if (SIZEOF_LONG != 4) || (SIZEOF_WCHAR_T != 2)
   uint64_entry:
 #endif
   case FFI_RETURN_TYPE_UINT64_T:
